File: easy/queue_using_stack/queue_using_stack.py
class Stack(object):
    def __init__(self):
        self.elements = []
        return

    def push(self, x):
        self.elements.append(x)
        return

    def pop(self):
        if self.empty():
            return 0
        return self.elements.pop()

    def peek(self):
        if self.empty():
            return 0
        return self.elements[len(self.elements)-1]

    def empty(self):
        return len(self.elements) == 0


class MyQueue(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.stack = Stack()
        return

    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: None
        """
        temp = Stack()
        while not self.stack.empty():
            element = self.stack.pop()
            temp.push(element)

        self.stack.push(x)
        while not temp.empty():
            element = temp.pop()
            self.stack.push(element)
        return

    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        return self.stack.pop()

    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        return self.stack.peek()

    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return self.stack.empty()


# Your MyQueue object will be instantiated and called as such:
obj = MyQueue()
obj.push(1)
obj.push(2)
a = obj.pop()
b = obj.peek()
c = obj.empty()

print("a = ", a, ", b = ", b, ", c = ", c)

File: easy/queue_using_stack/queue_using_stack.go
package main

import "fmt"

type Stack struct {
	elements []int
}

func NewStack() *Stack {
	return &Stack{
		elements: []int{},
	}
}

func (s *Stack) Push(x int) {
	s.elements = append(s.elements, x)
	return
}

func (s *Stack) Pop() int {
	if s.Empty() {
		return 0
	}

	top := s.elements[len(s.elements)-1]
	s.elements = s.elements[:len(s.elements)-1]
	return top
}

func (s *Stack) Peek() int {
	return s.elements[len(s.elements)-1]
}

func (s *Stack) Empty() bool {
	return len(s.elements) == 0
}

type MyQueue struct {
	stack *Stack
}

/** Initialize your data structure here. */
func Constructor() MyQueue {
	return MyQueue{
		stack: NewStack(),
	}
}

/** Push element x to the back of queue. */
func (q *MyQueue) Push(x int) {
	temp := NewStack()
	for !q.stack.Empty() {
		element := q.stack.Pop()
		temp.Push(element)
	}

	q.stack.Push(x)
	for !temp.Empty() {
		element := temp.Pop()
		q.stack.Push(element)
	}

	return
}

/** Removes the element from in front of queue and returns that element. */
func (q *MyQueue) Pop() int {
	return q.stack.Pop()
}

/** Get the front element. */
func (q *MyQueue) Peek() int {
	if q.Empty() {
		return 0
	}
	return q.stack.Peek()
}

/** Returns whether the queue is empty. */
func (q *MyQueue) Empty() bool {
	return q.stack.Empty()
}

func main() {
	stack := NewStack()
	stack.Push(1)
	stack.Push(2)
	stack.Push(3)
	a := stack.Peek()
	b := stack.Pop()
	c := stack.Empty()
	d := stack.Peek()
	fmt.Println("a = ", a, ", b = ", b, ", c = ", c, ", d = ", d)

	queue := Constructor()
	queue.Push(1)
	queue.Push(2)
	a = queue.Peek()
	b = queue.Pop()
	c = queue.Empty()
	fmt.Println("a = ", a, ", b = ", b, ", c = ", c)
}
